Add components ModelList and Pair

thallgren opened the pull request:
This commit adds two components that are useful when rendering models. A
`ModelList` and a `Pair`.

The `ModelList` is modelled after the `CollectionList` with some major
differences:

- It responds to model `change` events instead of `add` and `remove`.
- It expects a `componentFactory` function that takes a `key` and a
  `value` instead of an `item` and an `index`.

The default behavior is to call the `componentFactory` once for each
association in the model using alphabetical ordering of the keys. The
actual set of associations can be controlled using the options `include`
and `exclude`, both arrays of keys. The `include` option, if present,
overrides the alphabetical ordering. A key present in both `exclude` and
`include` will be excluded.

The `Pair` is just a grouping of two `Elem` instances and provided as a
convenient way to return a component from the component factory that
contains both a key and a value component.

thallgren committed:
Add components ModelList and Pair

This commit adds two components that are useful when rendering models. A
`ModelList` and a `Pair`.

The `ModelList` is modelled after the `CollectionList` with some major
differences:

- It responds to model `change` events instead of `add` and `remove`.
- It expects a `componentFactory` function that takes a `key` and a
  `value` instead of an `item` and an `index`.

The default behavior is to call the `componentFactory` once for each
association in the model using alphabetical ordering of the keys. The
actual set of associations can be controlled using the options `include`
and `exclude`, both arrays of keys. The `include` option, if present,
overrides the alphabetical ordering. A key present in both `exclude` and
`include` will be excluded.

The `Pair` is just a grouping of two `Elem` instances and provided as a
convenient way to return a component from the component factory that
contains both a key and a value component.

jirenius commented on src/Pair.js:
```diff
@@ -0,0 +1,28 @@
+import { Elem, RootElem } from 'modapp-base-component';
+
+// A component rendering a key,value pair
+class Pair extends RootElem {
```
The extended *RootElem* is never used for rendering. See review comments.

thallgren replied:
I'll change it and submit a new PR to modapp-base-components.

jirenius commented on src/ModelList.js:
```diff
@@ -0,0 +1,354 @@
+class ModelList extends RootElem {
+
+	/**
+     * Creates an instance of ModelList
```
Indentation: JSDoc comments indentation should be tab-indented and space-aligned (just like in Go :) ):  `<tab><space>` instead of `<space> * 5`.

jirenius commented on src/ModelList.js:
```diff
@@ -0,0 +1,354 @@
+	/**
+     * Sets the model.
+     * If the component is rendered, the list will be rerendered with
+     * the new collection, without any animation.
```
collection => model

jirenius reviewed (changes requested):
### ModelList
*ModelList* looks nice! I really like the exclude and include features. There are few minor things, but overall very nice.

* Search for the word `collection`, and most of the time you can replace it with `model` (including `this.collection`). Remnants from CollectionList.
* JSDoc comment indentation should be tab+space instead of just spaces.

### Pair
*Pair* solves a valid issue, but might need a bit of rethinking.

First, it suits the *modapp-base-component* repository better, as it has nothing to do with resources (models or collections). So it can be removed from this PR, and added to a PR to *modapp-base-component*. 

Secondly, the extended *RootElem* is currently not used:

```javascript
// Create a <dt><dd> pair
var pair = new Pair(
	n => n.elem('dt', [ n.text("Word") ]),
	n => n.elem('dd', [ n.text("Definition") ])
);
// Renders the two underlying Elem components, but not the RootElem.
pair.render(document.body);
// Calling RootElem methods does nothing, as the RootElem is not rendered.
pair.setClassName("foo");
```

And the underlying key/value components are not accessible except by accessing the internal `this.key` and `this.value` properties (in which case a getter might be better, and would allow better jsdoc comments).

Instead, I would say, make it a base class without extending anything. Have the *Pair* component just take 2 separate components instead of Elem nodes. And then preferably add a way to access the components:

```javascript
// A component rendering a key,value pair
class Pair {
	/**
	 * Creates a new Pair instance
	 * @param {Component} key Key component
	 * @param {Component} value Value component
	 */
	constructor(key, value) {
		this._key = key;
		this._value = value;
	}

	render(el) {
		this._key.render(el);
		this._value.render(el);
	}

	unrender() {
		this._key.unrender();
		this._value.unrender();
	}

	/**
	 * Key component.
	 * Maybe, instead of a getter, we could have getKeyComponent
	 */
	get key() {
		return this._key;
	}

	/**
	 * Value component.
	 * Maybe, instead of a getter, we could have getValueComponent
	 */
	get value() {
		return this._value;
	}
}

export default Pair;
```

Usage would be similar, but we can use any component, such as `Txt`:
```javascript
// Create a <dt><dd> pair
var pair = new Pair(
	new Txt("Word", { tagName: 'dt' }),
	new Txt("Definition", { tagName: 'dd' }),
);
pair.render(document.body);
// Access the key component to set a new text.
pair.key.setText("Foo");
```

thallgren committed:
Remove Pair.js and enforce correct tab indentation

This commit removes the Pair.js component (will be submitted as a
separate PR to modapp-base-component and fixes some inadvertent space
indentation in JSDoc comments.
